Finished Login User Story

BrendanHorst opened the pull request:
- Added login page
- If a user tries to submit a form with incomplete or incorrect data, they do not get logged in and receive a flashed message
- If a user who is not logged in tries to access the '/' route, they get redirected to '/login'
- A logged-in user can click the 'log out' link to clear their session

BrendanHorst committed:
Wrote tests for logging in

BrendanHorst committed:
Added login.html and registered the /login route

BrendanHorst committed:
Added the ability for users to log in

Also validates form submissions for incorrect or missing data

BrendanHorst committed:
Added login_required wrapper and /logout route

Index ('/') can now only be accessed while logged in

BrendanHorst committed:
Added login button to base.html

Also added some styles to the base template and comments

BrendanHorst committed:
Added some styles for the flashed messages on the login page

zachfedor commented on portal/auth.py:
```diff
@@ -0,0 +1,94 @@
+
+    if request.method == 'POST':
+
+        con = db.get_db()
+        cur = con.cursor()
+
+        email = request.form['email']
+        password = request.form['password']
+
+        cur.execute('SELECT * FROM users WHERE email = %s', (email,))
+        user = cur.fetchone()
```
Swap order of these steps:
1. get form data
2. open connection and cursor
3. run queries and fetch data
4. close connection and cursor just in case of errors
5. process form...

zachfedor commented on portal/auth.py:
```diff
@@ -0,0 +1,94 @@
+        error = None
+
+        #Check for empty email form
+        if not email:
+            print('no email')
```
Unnecessary print statements

zachfedor commented on portal/auth.py:
```diff
@@ -0,0 +1,94 @@
+
+            cur.close()
+            con.close()
+
+            return redirect('/')
```
Swap hard coded URL for `url_for()` function

zachfedor reviewed (commented):
Most of it looks good, just small cleanup things!

BrendanHorst committed:
Fixed some minor review issues

-Closed the database connection immediately after querying
-Added comments to test_auth.py
-Removed print statements

Thommond reviewed (approved):
I am satisfied with the styles and see no major problems with the code.
